### DISCUSSION TITLE
How to add a timestamp field with generation_time to a query rule with TUMBLINGWINDOW

### DISCUSSION THREAD

**Discussion by canob** [lf-edge/ekuiper#2354] [Q&A] [answered]
Hi everyone,
I'm trying to create a metric of number of DNS queries from each IP address, every 5 minutes, from DNS Server logs.
I'm using this query for the rule:
`SELECT source_ip AS source_ip, COUNT(source_ip) AS count_source_ip FROM dns where tag = "dns_query" GROUP BY source_ip, TUMBLINGWINDOW(mi, 5);`
And the output to a file is this one:
```
{"count_source_ip":9,"source_ip":"10.11.21.19"}
{"count_source_ip":9,"source_ip":"10.11.21.18"}
{"count_source_ip":9,"source_ip":"10.11.2.40"}
```
What I want is that the output to the file (or other sink output) has the timestamp of the generated metric event, like this:
```
{"timestamp": "2023-10-24 15:15:00.108","count_source_ip":9,"source_ip":"10.11.21.19"}
{"timestamp": "2023-10-24 15:15:00.108","count_source_ip":9,"source_ip":"10.11.21.18"}
{"timestamp": "2023-10-24 15:15:00.107","count_source_ip":9,"source_ip":"10.11.2.40"}
```
Is ok even if the timestamp is in other format (for example epoch), but I need it as part of the event, as a field.
Is that possible?
I don't know how to do it, because obviously the base events have a timestamp field, but I cannot aggregate and group over that field.
Thanks in advance for your help.

**Comment by ngjaying** [comment, marked as answer]
Use window_start() or window_end().  Please check https://ekuiper.org/docs/en/latest/sqls/functions/other_functions.html#window-start

**Reply by canob** [reply]
Thanks @ngjaying !
